### DISCUSSION TITLE
Performance Issue in Windows 11

### DISCUSSION THREAD

**Discussion by kdien** [wezterm/wezterm#5400] [Q&A]
### What Operating System(s) are you running on?

Windows

### WezTerm version

Latest nightly

### Ask your question!

I'm running WSL2 with tmux, on Windows 11 for work. 4K monitor with 200% scaling. Integrated Intel GPU.

The same [WezTerm config](https://github.com/kdien/dotfiles/blob/master/wezterm/wezterm.lua) works great on Linux and macOS, but on Windows it's just slow, as in input is not smooth, is a bit laggy with slight delays for every keystroke. Windows Terminal and Alacritty are fast, but I hate their font rendering. WezTerm is the only terminal where I get good and consistent font rendering, even on laptop screen with fractional scaling, so I'd love to be able to at least mitigate this issue. 

Has anyone run into this? I've tried some tuning in the docs, including stuff like fps and different `front_end` configs to no avail. Anything else I can try? Happy to provide more info where needed.

**Comment by kdien** [comment]
I've also removed the following lines and still no difference
```lua
config.window_background_opacity = 0
config.win32_system_backdrop = 'Mica'
```

**Comment by kdien** [comment]
Hmm I've just tried the latest Nightly build (20240509...) and it seems to be better, sitting about in between Windows Terminal and WezTerm stable. I prefer the font rendering enough to trade off some performance, so I'm sticking with nightly for now to see.

**Comment by kdien** [comment]
So I've been on latest nightly for a few days, and performance sadly is still quite choppy, with latency between typing and text appearing on screen. It seems to work better on laptop screen (1080p @ 125% scaling) than external monitor (4k @ 200% scaling). Not sure if this is any indication.

Looking for any input on how I could help to debug or mitigate the issue.

**Reply by Crypto-Spartan** [reply]
seconding this. currently experiencing noticeable lag on windows 11 on a 4k monitor, _plenty_ of CPU/RAM headroom to spare.

**Comment by clorl** [comment]
Same issue here, are there any logs that I could provide for this?

**Comment by joelhulander** [comment]
I have the same issue. When scrolling in neovim the CPU is considerably higher in WezTerm compared to Windows Terminal for example.

**Reply by sebastianstudniczek** [reply]
Yeah, me to, for some reason running Neovim in Wezterm on Windows is quite laggy, but on Windows Terminal it's smooth

**Reply by rpop0** [reply]
@sebastianstudniczek @joelhulander 
I really recommend using Wezterm via SSH on WSL. I switched to it and the performance is significantly better.

1.  Install openssh-server
2. enable the daemon
3.  create a ssh keypair on windows and drop the public key on WSL.
4.  Install Wezterm in WSL, make sure your version matches to the one installed on Windows.
5. In your wezterm config then do:
```
config.ssh_domains = {
  {
    name = 'WSL_SSH'',
    remote_address = 'localhost',
    username = 'yourusername',
  },
}
config.default_domain = 'WSL_SSH'
```
Opening wezterm now should just drop you in WSL and everything should be working as normal, but a bit faster. 

Just as a note, you will have to start your WSL distro separately on restart (Or configure it to start automatically with Windows), since unlike the normal WSL domain, this does not start it directly.

**Reply by sebastianstudniczek** [reply]
@rpop0 Thanks for sharing but actually I am expieriencing it with Neovim on Windows, not using currently it within WSL

**Reply by Crypto-Spartan** [reply]
i have the same issue as @sebastianstudniczek, very laggy with neovim on windows

**Reply by sebastianstudniczek** [reply]
Tried nightly build and it's better but still not as smooth as on Windows Terminal and have some problem with highlight groups in plain Vim. Changing gpu adapter didn't make difference

**Reply by Altamimi-Dev** [reply]
any update on this i am facing the same issue with neovim in windows 11

**Reply by sebastianstudniczek** [reply]
I gave up on Neovim on Windows since in general it's noticeably slower (mainly due to slow windows I/O system) then on WSL.

IMO if you're using Windows then Windows Terminal/Alacritty is the way to go in terms of performance

**Reply by kdien** [reply]
I've experienced some improvements lately with the latest nightly build and the following config updates:
```lua
config.max_fps = 120
config.animation_fps = 1
config.front_end = 'OpenGL'
config.webgpu_power_preference = 'HighPerformance'
config.prefer_egl = true
```

YMMV since I don't have my full blown Neovim setup on Windows, but WezTerm is now at least (for me) acceptable enough to make the much better font rendering worth it.

**Reply by sebastianstudniczek** [reply]
@kdien Out of curiosity, which font are you using? I am using JetBrains Mono NF but don't see that big of a difference in rendering compared to Windows Terminal

**Reply by kdien** [reply]
@sebastianstudniczek yeah not all fonts are noticeable. For me, the "affected" fonts are Meslo, DejaVu Sans Mono, and Berkeley Mono. WezTerm uses freetype, while Windows Terminal and Alacritty use the native DirectWrite, which makes the shaping of the fonts different and also makes fonts generally bolder. Doesn't help that I'm quite "sensitive" to fonts 😅

**Reply by sebastianstudniczek** [reply]
For me Alacritty is rendering font differently compared to Wezterm/Windows Terminal. Maybe it's a matter of width/height setting but for example mentioned JetBrains Mono is taller, thinner and don't have sharp edges when using small font size

**Reply by tfriedel** [reply]
I made a PR that massively reduces input lag on Windows / WSL:
https://github.com/wezterm/wezterm/pull/7470

I still want to test it a bit more, but please anyone who faced issues please let me know if it solves it for you.

**Reply by sebastianstudniczek** [reply]
@tfriedel Not sure whether this should help with such issue but I tested your version by opening neovim in tmux and was scrolling up down in larger file and unfortunately can't see big difference between this and nightly version. Checked also `AutoNoVsync` and `Immediate` with the same result

config:
```lua
config = {
	max_fps = 120,
	front_end = "WebGpu",
	webgpu_present_mode = "Mailbox",
	webgpu_max_frame_latency = 1
}
```

**Reply by tfriedel** [reply]
@sebastianstudniczek thanks for giving it a try!
I tried to replicate what you observed, i.e. laggy scrolling in neovim in tmux in wsl and I can see it too. I connected to the same tmux session twice, once with windows terminal and once with wezterm, each window taking up half of the monitor. I did observe some noticeable stuttering / unsmooth scrooling in wezterm and less of that in windows terminal. 

I gues there's a few different problems that need solving. I managed to get the input-latency down, but the unsmooth scrolling is a separate problem. Windows terminal unfortunately still has 32 ms input latency and since it's not open source there's nothing I can do about that.
I'll try to investigate if there's a way to fix this in wezterm when I have some time.

**Reply by sebastianstudniczek** [reply]
Nevertheless great work with reducing latency that much :D 
You made me aware of this problem, need to do some comparison with Alacritty.

On side note, AFAIK windows terminal is open source

**Reply by Crypto-Spartan** [reply]
Link to windows terminal repo: https://github.com/microsoft/terminal

**Reply by sebastianstudniczek** [reply]
I guess the differemce in performance with WSL might come from ConPTY, recently I stumbled upon Ghostty forks that are bypassing this layer and connecting directly into WSL

https://github.com/sudo-tee/hollow
https://github.com/Codavo/ghostinthewsl

I am using tmux + neovim within Windows Terminal and while at start it's pretty smooth, after some time it can get pretty laggy
